File: dynamic_programming/coin_change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    let memo = new Array(amount + 1).fill(-1);

    let dp = i => {
        if(i > amount){
            return Infinity;
        }
        if(i === amount){
            return 0;
        }
        if(memo[i] !== -1){
            return memo[i];
        }

        let min_amount = Infinity;
        for(let j = 0; j < coins.length; j++){
            let next = 1 + dp(i + coins[j]);
            min_amount = Math.min(min_amount, next);
        }

        memo[i] = min_amount;
        return memo[i];
    }

    let ans = dp(0);
    return ans === Infinity? -1 : ans;
};

let coins = [1,2,5];
let amount = 3;
let ans = coinChange(coins, amount);
console.log(ans);

File: data_structures_and_algorithms_crash_course/dynamic_programming/coin_change.py
from functools import cache
from typing import List

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        @cache
        def dp(i):
            if i > amount:
                return float("inf")
            if i == amount:
                return 0
            
            min_amount = float("inf")
            for coin in coins:
                next = 1 + dp(i + coin)
                min_amount = min(min_amount, next)

            return min_amount
        
        return dp(0) if dp(0) != float("inf") else -1

def main():
    runner = Solution()
    coins = [2]
    amount = 3
    ans = runner.coinChange(coins, amount)
    print(ans)

if __name__ == "__main__":
    main()